terraform-provider-constellation/internal/provider/convert.go

```diff
@@ -0,0 +1,173 @@
+			FirmwareSignerConfig: firmwareCfg,
+			AMDRootKey:           rootKey,
+		}
+	}
+	return attestationConfig, nil
+}
+
+func convertSNPAttestationTfStateCompatible(attestationVariant variant.Variant,
+	snpVersions attestationconfigapi.SEVSNPVersionAPI,
+) (tfSnpAttestation sevSnpAttestation, err error) {
+	var cert config.Certificate
+	switch attestationVariant.(type) {
+	case variant.AWSSEVSNP:
+		cert = config.DefaultForAWSSEVSNP().AMDRootKey
+	case variant.AzureSEVSNP:
+		cert = config.DefaultForAzureSEVSNP().AMDRootKey
+	}
+	certBytes, err := cert.MarshalJSON()
+	if err != nil {
+		return tfSnpAttestation, err
+	}
+	tfSnpAttestation = sevSnpAttestation{
+		BootloaderVersion: snpVersions.Bootloader,
+		TEEVersion:        snpVersions.TEE,
+		SNPVersion:        snpVersions.SNP,
+		MicrocodeVersion:  snpVersions.Microcode,
+		AMDRootKey:        string(certBytes),
+	}
+	if attestationVariant.Equal(variant.AzureSEVSNP{}) {
+		firmwareCfg := config.DefaultForAzureSEVSNP().FirmwareSignerConfig
+		tfFirmwareCfg, err := convertToTfSNPFirmwareSignerConfig(firmwareCfg)
+		if err != nil {
+			return tfSnpAttestation, err
+		}
+		tfSnpAttestation.AzureSNPFirmwareSignerConfig = tfFirmwareCfg
+	}
+	return tfSnpAttestation, nil
+}
+
+func convertToTfSNPFirmwareSignerConfig(firmwareCfg config.SNPFirmwareSignerConfig) (azureSnpFirmwareSignerConfig, error) {
+	keyDigestAny, err := firmwareCfg.AcceptedKeyDigests.MarshalYAML()
+	if err != nil {
+		return azureSnpFirmwareSignerConfig{}, err
+	}
+	keyDigest, ok := keyDigestAny.([]string)
+	if !ok {
+		return azureSnpFirmwareSignerConfig{}, errors.New("reading Accepted Key Digests: could not convert to []string")
+	}
+	return azureSnpFirmwareSignerConfig{
+		AcceptedKeyDigests: keyDigest,
+		EnforcementPolicy:  firmwareCfg.EnforcementPolicy.String(),
+		MAAURL:             firmwareCfg.MAAURL,
+	}, nil
+}
+
+func convertFromTfSNPFirmwareSignerConfig(tfFirmwareCfg azureSnpFirmwareSignerConfig) (config.SNPFirmwareSignerConfig, error) {
+	var keyDigests idkeydigest.List
+	if err := keyDigests.ParseStringSlice(tfFirmwareCfg.AcceptedKeyDigests); err != nil {
+		return config.SNPFirmwareSignerConfig{}, err
+	}
+	return config.SNPFirmwareSignerConfig{
+		AcceptedKeyDigests: keyDigests,
+		EnforcementPolicy:  idkeydigest.EnforcePolicyFromString(tfFirmwareCfg.EnforcementPolicy),
+		MAAURL:             tfFirmwareCfg.MAAURL,
+	}, nil
+}
+
+func convertMeasurementsTfStateCompatible(m measurements.M) map[string]measurement {
+	tfMeasurements := map[string]measurement{}
+	for key, value := range m {
+		keyStr := strconv.FormatUint(uint64(key), 10)
+		tfMeasurements[keyStr] = measurement{
+			Expected: hex.EncodeToString(value.Expected),
+			WarnOnly: bool(value.ValidationOpt),
+		}
+	}
+	return tfMeasurements
+}
+
+type extraMicroservices struct {
+	CSIDriver bool `tfsdk:"csi_driver"`
+}
+
+type measurement struct {
+	Expected string `tfsdk:"expected"`
+	WarnOnly bool   `tfsdk:"warn_only"`
+}
+
+type sevSnpAttestation struct {
```

Comment:
Don't we need to use Terraform's `types` package here?

Reply:
no, thats the magic of the type annotation :)

Reply:
Possibly we could directly assign these tags to the types we implement in internal then and spare the duplication?

Reply:
agreed to keep for now, as we don't want to couple the constellation API to the TF provider. Might be reworked later.